File: verdict-backend/src/routes/cases.ts
import type { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
import { z } from 'zod';
import { db } from '../lib/db.js';
import { requireAuth } from '../middleware/auth.js';

const createCaseBody = z.object({
  name: z.string().min(3).max(255).trim(),
  caseType: z.enum([
    'MEDICAL_MALPRACTICE',
    'EMPLOYMENT_DISCRIMINATION',
    'COMMERCIAL_DISPUTE',
    'CONTRACT_BREACH',
    'OTHER',
  ]),
  caseTypeCustom: z.string().max(255).optional(),
  opposingFirm: z.string().max(255).optional(),
  depositionDate: z.string().datetime().optional(),
});

export async function casesRoutes(app: FastifyInstance) {
  app.addHook('preHandler', requireAuth);

  app.get('/', async (request: FastifyRequest, reply: FastifyReply) => {
    const user = request.user!;
    const cases = await db.case.findMany({
      where: { firmId: user.firmId, isArchived: false },
      orderBy: { depositionDate: 'asc' },
      include: {
        _count: { select: { documents: true, witnesses: true } },
      },
    });
    return reply.send({
      success: true,
      data: {
        cases: cases.map((c) => {
          const count = (c as { _count: { documents: number; witnesses: number } })._count;
          return {
            id: c.id,
            name: c.name,
            caseType: c.caseType,
            depositionDate: c.depositionDate,
            documentsReady: count?.documents ?? 0,
            documentsTotal: count?.documents ?? 0,
            witnessCount: count?.witnesses ?? 0,
            createdAt: c.createdAt,
          };
        }),
        pagination: { page: 1, limit: 20, total: cases.length },
      },
    });
  });

  app.post('/', async (request: FastifyRequest, reply: FastifyReply) => {
    const user = request.user!;
    const parsed = createCaseBody.safeParse(request.body);
    if (!parsed.success) {
      return reply.code(400).send({
        success: false,
        error: { code: 'VALIDATION_ERROR', message: 'Invalid body', details: parsed.error.flatten() },
      });
    }
    const { name, caseType, caseTypeCustom, opposingFirm, depositionDate } = parsed.data;
    const c = await db.case.create({
      data: {
        firmId: user.firmId,
        ownerId: user.id,
        name,
        caseType,
        caseTypeCustom,
        opposingFirm,
        depositionDate: depositionDate ? new Date(depositionDate) : null,
      },
    });
    return reply.code(201).send({
      success: true,
      data: {
        id: c.id,
        name: c.name,
        caseType: c.caseType,
        depositionDate: c.depositionDate,
        createdAt: c.createdAt,
      },
    });
  });

  app.get<{ Params: { caseId: string } }>('/:caseId', async (request, reply) => {
    const user = request.user!;
    const { caseId } = request.params;
    const c = await db.case.findFirst({
      where: { id: caseId, firmId: user.firmId },
      include: { documents: true, witnesses: true },
    });
    if (!c) return reply.code(404).send({ success: false, error: { code: 'NOT_FOUND', statusCode: 404 } });
    return reply.send({ success: true, data: c });
  });

  app.patch<{ Params: { caseId: string } }>('/:caseId', async (request, reply) => {
    const user = request.user!;
    const { caseId } = request.params;
    const body = request.body as Record<string, unknown>;
    const existing = await db.case.findFirst({
      where: { id: caseId, firmId: user.firmId },
    });
    if (!existing) return reply.code(404).send({ success: false, error: { code: 'NOT_FOUND', statusCode: 404 } });
    const updated = await db.case.update({
      where: { id: caseId },
      data: {
        ...(body.name != null && { name: String(body.name) }),
        ...(body.depositionDate != null && { depositionDate: new Date(String(body.depositionDate)) }),
      },
    });
    return reply.send({ success: true, data: updated });
  });

  app.delete<{ Params: { caseId: string } }>('/:caseId', async (request, reply) => {
    const user = request.user!;
    const { caseId } = request.params;
    const existing = await db.case.findFirst({
      where: { id: caseId, firmId: user.firmId },
    });
    if (!existing) return reply.code(404).send({ success: false, error: { code: 'NOT_FOUND', statusCode: 404 } });
    await db.case.update({ where: { id: caseId }, data: { isArchived: true } });
    return reply.send({ success: true, message: 'Case archived' });
  });
}

File: verdict-backend/app/routers/cases.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.database import get_db
from app.middleware.auth import require_auth
from app.models.user import User
from app.models.case import Case
from app.schemas.cases import CreateCaseRequest

router = APIRouter()


@router.get("/")
async def list_cases(db: AsyncSession = Depends(get_db), user: User = Depends(require_auth)):
    result = await db.execute(
        select(Case)
        .where(Case.firm_id == user.firm_id, Case.is_archived == False)
        .order_by(Case.deposition_date.asc())
    )
    cases = result.scalars().all()
    return {
        "success": True,
        "data": {
            "cases": [
                {
                    "id": c.id,
                    "name": c.name,
                    "caseType": c.case_type,
                    "depositionDate": c.deposition_date,
                    "createdAt": c.created_at,
                }
                for c in cases
            ],
            "pagination": {"page": 1, "limit": 20, "total": len(cases)},
        },
    }


@router.post("/", status_code=201)
async def create_case(body: CreateCaseRequest, db: AsyncSession = Depends(get_db), user: User = Depends(require_auth)):
    c = Case(
        firm_id=user.firm_id,
        owner_id=user.id,
        name=body.name,
        case_type=body.caseType,
        case_type_custom=body.caseTypeCustom,
        opposing_firm=body.opposingFirm,
        deposition_date=datetime.fromisoformat(body.depositionDate) if body.depositionDate else None,
    )
    db.add(c)
    await db.commit()
    await db.refresh(c)
    return {
        "success": True,
        "data": {"id": c.id, "name": c.name, "caseType": c.case_type, "depositionDate": c.deposition_date, "createdAt": c.created_at},
    }


@router.get("/{case_id}")
async def get_case(case_id: str, db: AsyncSession = Depends(get_db), user: User = Depends(require_auth)):
    result = await db.execute(select(Case).where(Case.id == case_id, Case.firm_id == user.firm_id))
    c = result.scalar_one_or_none()
    if not c:
        raise HTTPException(404, detail={"code": "NOT_FOUND"})
    return {"success": True, "data": {"id": c.id, "name": c.name, "caseType": c.case_type, "depositionDate": c.deposition_date}}


@router.patch("/{case_id}")
async def update_case(case_id: str, body: dict, db: AsyncSession = Depends(get_db), user: User = Depends(require_auth)):
    result = await db.execute(select(Case).where(Case.id == case_id, Case.firm_id == user.firm_id))
    c = result.scalar_one_or_none()
    if not c:
        raise HTTPException(404, detail={"code": "NOT_FOUND"})
    if "name" in body:
        c.name = body["name"]
    if "depositionDate" in body:
        c.deposition_date = datetime.fromisoformat(body["depositionDate"])
    await db.commit()
    return {"success": True, "data": {"id": c.id, "name": c.name}}


@router.delete("/{case_id}")
async def archive_case(case_id: str, db: AsyncSession = Depends(get_db), user: User = Depends(require_auth)):
    result = await db.execute(select(Case).where(Case.id == case_id, Case.firm_id == user.firm_id))
    c = result.scalar_one_or_none()
    if not c:
        raise HTTPException(404, detail={"code": "NOT_FOUND"})
    c.is_archived = True
    await db.commit()
    return {"success": True, "message": "Case archived"}
